### PR TITLE
Add configurable soft limiting of memory used by merge operations on a content node

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim @geirst please review

This adds heuristically derived estimates of memory footprint to all merge operations sent from distributor nodes, and tracking of potential memory usage in the `MergeThrottler` component on the content nodes. If an incoming merge operation would exceed the configured limits, it is busy-bounced (or potentially enqueued, if coming directly from a distributor).

This also wires in `HwInfo` to the service layer setup code, which is used when auto-deducing memory limits based on config and available node memory.

For now, memory limiting is disabled by default. Testing and roll-out will be done gradually—it is expected that auto-deduction will be the default in the end.

### CODE REVIEW

**Review by baldersheim — DISMISSED** [review]

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp [hunk 1]
```diff
@@ -761,7 +791,10 @@ MergeThrottler::handleMessageDown(
 
         if (isMergeAlreadyKnown(msg)) {
             processCycledMergeCommand(msg, msgGuard);
-        } else if (canProcessNewMerge() || allow_merge_despite_full_window(mergeCmd)) {
+        } else if (accepting_merge_is_within_memory_limits(mergeCmd)
+                   && (canProcessNewMerge()
+                       || allow_merge_despite_full_window(mergeCmd)))
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/vespa/storage/storageserver/mergethrottler.cpp:801]: `allow_merge_despite_full_window` is only checked iff `accepting_merge_is_within_memory_limits` is already true, so it should not override its decisions. Unless there's something else I'm missing?

File: storage/src/vespa/storage/distributor/operations/idealstate/mergeoperation.cpp [hunk 2]
```diff
@@ -367,6 +371,40 @@ bool MergeOperation::all_involved_nodes_support_unordered_merge_chaining() const
     return true;
 }
 
+uint32_t MergeOperation::estimate_merge_memory_footprint_ubound(const std::vector<MergeMetaData>& nodes) const noexcept {
```
[the remark this answers is not in the document]
Comment (vekterli) [reply storage/src/vespa/storage/distributor/operations/idealstate/mergeoperation.cpp]: Renamed

[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operations/idealstate/mergeoperation.cpp]: 👍

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/storageserver/mergethrottler.cpp:801]: So intentional then ..., good. 👍

**Review by baldersheim — APPROVED** [review]

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp [hunk 3]
```diff
@@ -65,6 +67,8 @@ MergeThrottler::Metrics::Metrics(metrics::MetricSet* owner)
       averageQueueWaitingTime("averagequeuewaitingtime", {}, "Average time a merge spends in the throttler queue", this),
       queueSize("queuesize", {}, "Length of merge queue", this),
       active_window_size("active_window_size", {}, "Number of merges active within the pending window size", this),
+      estimated_merge_memory_usage("estimated_merge_memory_usage", {}, "An estimated upper bound of the "
```
Comment (geirst) [inline storage/src/vespa/storage/storageserver/mergethrottler.cpp:70]: Please add in the comment that this is in bytes.

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp [hunk 4]
```diff
@@ -69,6 +69,7 @@ MergeThrottler::Metrics::Metrics(metrics::MetricSet* owner)
       active_window_size("active_window_size", {}, "Number of merges active within the pending window size", this),
       estimated_merge_memory_usage("estimated_merge_memory_usage", {}, "An estimated upper bound of the "
                                    "memory usage of the merges currently in the active window", this),
+      merge_memory_limit("merge_memory_limit", {}, "The active soft limit for memory used by merge operations on this node", this),
```
Comment (geirst) [inline storage/src/vespa/storage/storageserver/mergethrottler.cpp:72]: Please add in the comment that this is in bytes.

**Review by geirst — COMMENTED** [review]
👍

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp (hunk 3 above)
Comment (vekterli) [reply storage/src/vespa/storage/storageserver/mergethrottler.cpp:70]: Addressed in #29205

File: storage/src/vespa/storage/storageserver/mergethrottler.cpp (hunk 4 above)
Comment (vekterli) [reply storage/src/vespa/storage/storageserver/mergethrottler.cpp:72]: Addressed in #29205

**Pull request merged**